src/uct/ib/base/ib_iface.c

```diff
@@ -985,6 +1007,12 @@ ucs_status_t uct_ib_iface_create_qp(uct_ib_iface_t *iface,
                   attr->cap.max_inline_data, attr->max_inl_cqe[UCT_IB_DIR_TX],
                   attr->cap.max_recv_wr, attr->cap.max_recv_sge,
                   attr->max_inl_cqe[UCT_IB_DIR_RX]);
+
+        /* Verbs have no way to detect max_inline_data, only to try and fail */
+        if (attr->cap.max_inline_data < attr->ibv.cap.max_inline_data) {
+            attr->cap.max_inline_data = attr->ibv.cap.max_inline_data;
+            goto retry_qp_creation;
```

Comment:
can't it be an infinite loop?

Reply:
I don't see how it could: if we set max_inline_data to X and `ibv_create_qp_*` returns lower than X - then we redo (for >=X we don't). Verbs shouldn't give us a different value for max_inline_data after every call. In the unlikely fault in verbs - the worst that can happen is we decrement after every invocation until max_inline_data is 0.

Reply:
1. attr->ibv.cap.max_inline_data is what ibv_create_qp returns since we do `ibv_create_qp_ex(dev->ibv_context, &attr->ibv);`
2. attr->cap.max_inline_data is what was configured by the user and set in uct_ud_iface_create_qp
3. according to ibv_create_qp man page, it returns cap which are larger or equal to passed values if qp create is success, and there is no definition of what caps will be in case of failure

Reply:
1 + 2 - yes, the code intends to use this. I saw [Dotan Barak's comment ](https://www.rdmamojo.com/2012/12/21/ibv_create_qp/#comment-233) suggesting ibv.max_inline_data is set in case of a failure (otherwise there's no way to detect how much inline data is supported), but I can understand if you don't want to rely on it...

As an alternative, I suggest retrying one more time with max_inline_data=0, to make sure we support NICs with a smaller inline data support than UCX's default configuration.

Reply:
correction, I meant [this comment by Dotan Barak](https://www.rdmamojo.com/2013/01/26/ibv_post_send/#comment-411)

Reply:
That comment is when QP create is successful. The prev page says:
"In some RDMA devices, creating a QP with will set the value of max_inline_data to the size of messages that can be sent using the requested number of scatter/gather elements of the Send Queue. If others, one should specify explicitly the message size to be sent inline before the creation of a QP. for those devices, it is advised to try to create the QP with the required message size and continue decreasing it if the QP creation fails."

so - retry one more time with 0